### PR TITLE
Unique validators

### PR DESCRIPTION
**Pull request by wpdebruin**
We should think about ways to test the db and orm stuff.

I added the orm validation conditionally, by checking if cborm was loaded and if we had an entity.

If not, you have to configure at least a table property, and in some cases some more properties. It should work for both model and struct validation, but when using for structs you have to include the id (key)

### CODE REVIEW

**Comment by lmajano** [comment]
@wpdebruin The cborm module already has the unique validation in the module. This validator should be a vanilla one for any db without considering orm.

File: models/validators/UniqueValidator.cfc [hunk 1]
```diff
@@ -2,64 +2,113 @@
...
+	property name="res" 				inject="model:resourceService@cbi18n";
+    property name="moduleService" inject="coldbox:moduleService";
+    property name="wirebox" inject="wirebox";
 
 	/**
 	 * Constructor
 	 */
-	UniqueValidator function init(){
-		variables.name = "Unique";
+	QUniqueValidator function init() {
```
Comment (lmajano) [inline models/validators/UniqueValidator.cfc:17]: Why `QUniqueValidator`?

File: models/validators/UniqueValidator.cfc [hunk 2]
```diff
@@ -2,64 +2,113 @@
...
-			"SELECT 1 FROM #arguments.validationData.table# WHERE #targetColumn# = ?",
-			[ arguments.targetValue ]
-		).recordCount > 0;
+	) {
 
-		if ( !exists ) {
-			return true;
-		}
+        //check to see if cborm is loaded and if this is a cform entity
+        if ( moduleService.isModuleActive("cborm") and getMetadata(arguments.target).persistent ?: false ) {
```
Comment (lmajano) [inline models/validators/UniqueValidator.cfc:39]: let's not worry about cborm. cborm has it's own validator.

File: models/validators/UniqueValidator.cfc [hunk 3]
```diff
@@ -2,64 +2,113 @@
...
+				.newError( argumentCollection = args )
+				.setErrorMetadata( { QUniqueValidator : arguments.validationData } )
+		);
 		return false;
 	}
 
+	/**
+	 * Get the name of the validator
+	 */
+	string function getName() {
```
Comment (lmajano) [inline models/validators/UniqueValidator.cfc:110]: This should come from the abstract one

File: models/validators/UniqueValidator.cfc [hunk 4]
```diff
@@ -2,64 +2,113 @@
...
- * - table : The table name to seek
- * - column : The column to evaluate for uniqueness or defaults to the name of the field
+ * This validator validates if a value is unique in a database
  */
-component
-	extends  ="BaseValidator"
-	accessors="true"
-	singleton
-{
+component accessors="true" singleton {
```
Comment (lmajano) [inline models/validators/UniqueValidator.cfc:7]: This should extend the base one

File: models/validators/UniqueValidator.cfc [hunk 5]
```diff
@@ -2,64 +2,113 @@
...
-					validationData : arguments.validationData
-				}
-			)
-		);
+            // return true if no data to check, type needs a data element to be checked.
+            if ( isNullOrEmpty( arguments.targetValue ) ) {
+                return true;
+            }
 
+            var sql     = "Select 1 from #validationData.table# where #validationData.column# = :name";
```
Comment (lmajano) [inline models/validators/UniqueValidator.cfc:70]: Just make sure the formatter runs

**Review by lmajano — CHANGES_REQUESTED** [review]

File: models/validators/UniqueValidator.cfc (hunk 1 above)
Comment (wpdebruin) [reply models/validators/UniqueValidator.cfc:17]: yes, stupid, copy-paste error. I had two different unique validators in my code, one based on qb, the other on queries. One of them had q in it, but we don't need it anymore. Only query based validator here

File: models/validators/UniqueValidator.cfc (hunk 2 above)
Comment (wpdebruin) [reply models/validators/UniqueValidator.cfc:39]: Ok, removed

File: models/validators/UniqueValidator.cfc (hunk 3 above)
Comment (wpdebruin) [reply models/validators/UniqueValidator.cfc:110]: removed

**Comment by lmajano** [comment]
Great job @wpdebruin it just needs docs now!

**Pull request merged**